pkg/controllers/deprovisioning/controller.go

```diff
@@ -0,0 +1,517 @@
+	// Given candidate nodes, compute best deprovisioning action
+	attempts := 0
+	var cmd deprovisioningCommand
+	var err error
+	success := false
+	// Each attempt will try at least one node, limit to that many attempts.
+	for attempts < len(nodes) {
+		cmd, err = d.computeCommand(ctx, attempts, nodes...)
+		if err != nil {
+			return ResultFailed, err
+		}
+		logging.FromContext(ctx).Debugf("This is the time of command creation, %s for deprovisioner %s", cmd.created, d.string())
+		logging.FromContext(ctx).Debugf("This is the time before validation, %s", c.clock.Now())
+		// Validate the command with a TTL
+		ok, err := c.validateCommand(ctx, cmd, d)
+		if err != nil {
+			return ResultFailed, fmt.Errorf("validating command, %w", err)
+		}
+		logging.FromContext(ctx).Debugf("This is the time after validation, %s", c.clock.Now())
+		// If validation succeeds, execute the command. Otherwise, we need to compute another command and
+		// try the rest of the candidate nodes. If any of the remaining candidateNodes are still valid
+		// for an action, continue with deprovisioning, even though we might not be taking the most optimal decision.
+		// This is meant to stop an issue if another controller in the cluster coincidentally creates churn at the same interval.
+		if ok {
+			success = true
+			break
+		}
+		attempts++
+	}
+	// Validation for all candidateNodes failed, so nothing can be done for this deprovisioner.
+	if !success {
+		return ResultNothingToDo, nil
+	}
+	// Execute the command
+	result, err := c.executeCommand(ctx, cmd)
+	if err != nil {
+		return ResultFailed, err
+	}
+	return result, nil
+}
+
+func (c *Controller) validateCommand(ctx context.Context, cmd deprovisioningCommand, d deprovisioner) (bool, error) {
+	// Only continue with the rest of the logic if the command has an executable action for the deprovisioner
+	if !d.isExecutableCommand(cmd) {
+		return false, nil
+	}
+	// deprovisioningTTL is how long we wait before re-validating our lifecycle command
+	remainingDelay := d.getTTL() - c.clock.Since(cmd.created)
+	if remainingDelay > 0 {
+		select {
+		case <-ctx.Done():
+			return false, fmt.Errorf("context canceled during %s validation", d.string())
+		case <-c.clock.After(remainingDelay):
+		}
+	}
+
+	candidateNodes, err := c.candidateNodes(ctx, d.shouldNotBeDeprovisioned)
+	if err != nil {
+		return false, fmt.Errorf("determining candidate node, %w", err)
+	}
+	nodes := c.mapNodes(cmd.nodesToRemove, candidateNodes)
+
+	// None of the chosen candidate nodes are valid for execution, so retry
+	if len(nodes) == 0 {
+		return false, nil
+	}
+
+	ok, err := d.validateCommand(ctx, nodes, cmd)
+	if err != nil {
+		return false, fmt.Errorf("determining candidate node, %w", err)
+	}
+	if !ok {
+		return false, nil
+	}
+	return true, nil
+}
+
+// candidateNodes returns nodes that appear to be currently deprovisionable based off of their provisioner
+// nolint:gocyclo
```

Comment:
Is there something special about the controller that requires us to turn gocyclo off? Should we turn gocyclo off generally?